lib/spack/spack/concretize.py

```diff
@@ -203,46 +202,91 @@ def concretize_version(self, spec):
+        if not architecture.platform_os:
+            architecture.platform_os = default_os
+
+        return True  # spec changed
+
     def concretize_architecture(self, spec):
-        """If the spec is empty provide the defaults of the platform. If the
-        architecture is not a string type, then check if either the platform,
-        target or operating system are concretized. If any of the fields are
-        changed then return True. If everything is concretized (i.e the
-        architecture attribute is a namedtuple of classes) then return False.
-        If the target is a string type, then convert the string into a
-        concretized architecture. If it has no architecture and the root of the
-        DAG has an architecture, then use the root otherwise use the defaults
-        on the platform.
-        """
-        root_arch = spec.root.architecture
+        """If the spec's architecture is already concrete then return False.
+        If the spec is missing the architecture portion of the spec then look
+        through the DAG and find a spec with an architecture. If such a spec
+        exists then assign it to the current spec, otherwise provide the
+        defaults for the system. If the architecture is a Arch object then
+        check if either the platform, target, or operating system are
+        concretized. If any of the fields are changed return True."""
         sys_arch = spack.spec.ArchSpec(spack.architecture.sys_type())
-        spec_changed = False
+        front_arch = spack.spec.ArchSpec(
+                spack.architecture.front_end_sys_type())
+
+        if spec.architecture and spec.architecture.concrete:
+            return False
 
-        if spec.architecture is None:
+        spec_changed = False
+        other_arch_spec = find_spec(spec, lambda x: x.architecture, spec.root)
+        other_arch = other_arch_spec.architecture
+        if not other_arch:
             spec.architecture = spack.spec.ArchSpec(sys_arch)
+        elif not spec.architecture and other_arch:
+            spec.architecture = other_arch
             spec_changed = True
 
-        default_archs = list(x for x in [root_arch, sys_arch] if x)
-        for arch in default_archs:
-            if spec.architecture.concrete:
-                break
+        platform = spack.architecture.get_platform(spec.architecture.platform)
 
-            replacement_fields = [k for k, v in iteritems(arch.to_cmp_dict())
-                                  if v and not getattr(spec.architecture, k)]
-            for field in replacement_fields:
-                setattr(spec.architecture, field, getattr(arch, field))
-                spec_changed = True
+        spec_changed = self._concretize_arch_component(platform, spec)
 
         if not spec.architecture.concrete:
-            raise InsufficientArchitectureInfoError(spec, default_archs)
-
+            raise InsufficientArchitectureInfoError(spec, [sys_arch,
```

Comment:
`sys_arch` and `front_arch` may be the same on some systems. I think it would make sense to check for that in the initializer for `InsufficientArchitectureInfoError` and only print one of them if they are the same.

(minor) Also I'd prefer line wrapping like

```
InsufficientArchitectureInfoError(
    spec, [sys_arch, front_arch])
```

or 


```
InsufficientArchitectureInfoError(
    spec,
    [sys_arch, front_arch])
```